src/Engine/ProtoCore/ProcedureTable.cs

```diff
@@ -392,30 +373,92 @@ public int IndexOf(string name, List<ProtoCore.Type> argTypeList, bool isStaticO
         /// <param name="functionName"></param>
         /// <param name="parameterTypes"></param>
         /// <returns></returns>
+        [Obsolete("This method will be removed in Dynamo 3.0. Please use GetFunctionBySignature with ProcedureMatchOptions instead.")]
         public ProcedureNode GetFunctionBySignature(string functionName, List<Type> parameterTypes)
         {
-            foreach (var f in Procedures)
+            GetFunctionBySignature(new ProcedureMatchOptions() 
+            { 
+                FunctionName = functionName, 
+                ParameterTypes = parameterTypes 
+            }, out ProcedureNode output);
+            return output;
+        }
+
+        /// <summary>
+        /// Get function by its signature.
+        /// </summary>
+        /// <param name="opts">Matching options</param>
+        /// <param name="outputProcNode">Output procedure node. Null if nothing is found</param>
+        /// <returns>Index of the ProcedureNode in the Procedures list. Returns -1 If nothing is found</returns>
+        public int GetFunctionBySignature(ProcedureMatchOptions opts, out ProcedureNode outputProcNode)
+        {
+            int outputProcNodeIndex = Constants.kInvalidIndex;
+            outputProcNode = null;
+
+            // how many default parameters are used
+            int smallestDefaultArgNum = int.MaxValue;
+            for (int ii = 0; ii < Procedures.Count; ++ii)
```

Comment:
just thinking - couldn't this be quite slow? Do we have a map of functions so we can lookup the function by name first, and then do the other checks?

Reply:
The refactored code should not be any slower than it was before.
I think, in most of code I refactored, that Dynamo first locates the class Node, then looks through the procedure list. So it should not be that slow.